File: src/TurboVision/Objects/Point.cs
using System.Runtime.InteropServices;

namespace TurboVision.Objects;

public unsafe partial class Point(void* ptr, bool owned, bool placement)
    : NativeObject<Point>(ptr, owned, placement)
{
    private sealed class Factory : NativeObjectFactory<Factory>
    {
        public Factory()
            : base(
                NativeMethods.TV_Point_placementSize,
                NativeMethods.TV_Point_placementNew,
                NativeMethods.TV_Point_new
            ) { }
    }

    public static int PlacementSize => Factory.Instance.PlacementSize;

    public Point(byte* placement)
        : this(Factory.Instance.PlacementNew(placement), owned: true, placement: true) { }

    public Point()
        : this(Factory.Instance.New(), owned: true, placement: false) { }

    protected override void PlacementDeleteCore(void* ptr)
    {
        TurboVisionException.Check(NativeMethods.TV_Point_placementDelete(ptr));
    }

    protected override void DeleteCore()
    {
        TurboVisionException.Check(NativeMethods.TV_Point_delete(Ptr));
    }

    protected override bool EqualsCore(Point other)
    {
        TurboVisionException.Check(NativeMethods.TV_Point_equals(Ptr, other.Ptr, out var equals));
        return equals;
    }

    protected override int GetHashCodeCore()
    {
        TurboVisionException.Check(NativeMethods.TV_Point_hash(Ptr, out var hash));
        return hash;
    }

    public void Add(Point adder)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        ObjectDisposedException.ThrowIf(adder.IsDisposed, adder);
        TurboVisionException.Check(NativeMethods.TV_Point_operator_add_in_place(Ptr, adder.Ptr));
    }

    public void Subtract(Point subber)
    {
        ObjectDisposedException.ThrowIf(IsDisposed, this);
        ObjectDisposedException.ThrowIf(subber.IsDisposed, subber);
        TurboVisionException.Check(
            NativeMethods.TV_Point_operator_subtract_in_place(Ptr, subber.Ptr)
        );
    }

    public static void Add(Point one, Point two, Point dst)
    {
        ObjectDisposedException.ThrowIf(one.IsDisposed, one);
        ObjectDisposedException.ThrowIf(two.IsDisposed, two);
        ObjectDisposedException.ThrowIf(dst.IsDisposed, dst);
        TurboVisionException.Check(NativeMethods.TV_Point_operator_add(one.Ptr, two.Ptr, dst.Ptr));
    }

    public static void Subtract(Point one, Point two, Point dst)
    {
        ObjectDisposedException.ThrowIf(one.IsDisposed, one);
        ObjectDisposedException.ThrowIf(two.IsDisposed, two);
        ObjectDisposedException.ThrowIf(dst.IsDisposed, dst);
        TurboVisionException.Check(
            NativeMethods.TV_Point_operator_subtract(one.Ptr, two.Ptr, dst.Ptr)
        );
    }

    public int X
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_Point_get_x(Ptr, out var x));
            return x;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_Point_set_x(Ptr, value));
        }
    }

    public int Y
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_Point_get_y(Ptr, out var y));
            return y;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_Point_set_y(Ptr, value));
        }
    }

    internal static partial class NativeMethods
    {
        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_placementSize(out int outSize, out int outAlignment);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_placementNew(byte* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_placementDelete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_new(out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_delete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_equals(
            void* self,
            void* other,
            [MarshalAs(UnmanagedType.I4)] out bool @out
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_hash(void* self, out int @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_operator_add_in_place(void* self, void* adder);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_operator_subtract_in_place(void* self, void* subber);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_operator_add(void* one, void* two, void* dst);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_operator_subtract(void* one, void* two, void* dst);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_get_x(void* self, out int @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_set_x(void* self, int x);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_get_y(void* self, out int @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_Point_set_y(void* self, int y);
    }
}

File: src/tvision4c/Objects/Point.cpp
#include "Point.h"

EXPORT tv::Error TV_Point_placementSize(int32_t* outSize, int32_t* outAlignment) {
    return tv::checkedSize<TPoint>(outSize, outAlignment);
}

EXPORT tv::Error TV_Point_placementNew(TPoint* self) {
    return tv::checkedPlacementNew(self);
}

EXPORT tv::Error TV_Point_placementDelete(TPoint* self) {
    return tv::checkedPlacementDelete(self);
}

EXPORT tv::Error TV_Point_new(TPoint** out) {
    return tv::checkedNew(out);
}

EXPORT tv::Error TV_Point_delete(TPoint* self) {
    return tv::checkedDelete(self);
}

EXPORT tv::Error TV_Point_equals(TPoint* self, TPoint* other, BOOL* out) {
    if (!out) {
        return tv::Error_ArgumentNull;
    }

    if (!self && !other) {
        *out = TRUE;
        return tv::Success;
    }

    if (!self || !other) {
        *out = FALSE;
        return tv::Success;
    }

    if (*self == *other) {
        *out = TRUE;
        return tv::Success;
    }

    *out = FALSE;
    return tv::Success;
}

EXPORT tv::Error TV_Point_hash(TPoint* self, int32_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    int32_t result = 0;
    tv::hash(self->x, &result);
    tv::hash(self->y, &result);

    *out = result;
    return tv::Success;
}

EXPORT tv::Error TV_Point_operator_add_in_place(TPoint* self, TPoint* adder) {
    if (!self || !adder) {
        return tv::Error_ArgumentNull;
    }

    *self += *adder;
    return tv::Success;
}

EXPORT tv::Error TV_Point_operator_subtract_in_place(TPoint* self, TPoint* subber) {
    if (!self || !subber) {
        return tv::Error_ArgumentNull;
    }

    *self -= *subber;
    return tv::Success;
}

EXPORT tv::Error TV_Point_operator_add(TPoint* one, TPoint* two, TPoint* dst) {
    if (!one || !two || !dst) {
        return tv::Error_ArgumentNull;
    }

    *dst = *one + *two;
    return tv::Success;
}

EXPORT tv::Error TV_Point_operator_subtract(TPoint* one, TPoint* two, TPoint* dst) {
    if (!one || !two || !dst) {
        return tv::Error_ArgumentNull;
    }

    *dst = *one - *two;
    return tv::Success;
}

EXPORT tv::Error TV_Point_get_x(TPoint* self, int32_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->x;
    return tv::Success;
}

EXPORT tv::Error TV_Point_set_x(TPoint* self, int32_t x) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->x = x;
    return tv::Success;
}

EXPORT tv::Error TV_Point_get_y(TPoint* self, int32_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->y;
    return tv::Success;
}

EXPORT tv::Error TV_Point_set_y(TPoint* self, int32_t y) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->y = y;
    return tv::Success;
}
